### PR TITLE
PLAT-98187 [Auto Agate] Add Dropdown component

### PR DESCRIPTION
**Pull request by RenukaAtale**
Added dropdown component to Agate

### CODE REVIEW

File: Dropdown/Dropdown.js [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+	},
+
+	styles: {
+		css: componentCss,
+		className: 'Dropdown',
+		publicClassNames: true
+	},
+
+	computed: {
+		className: ({css, selected, styler}) => styler.append(selected && css.selected),
```
Comment (teckliew) [inline Dropdown/Dropdown.js]: I don't think we have a `selected` style. Although maybe we should have a selected style for the selected option in the dropdown list.

File: Dropdown/Dropdown.js [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+	},
+
+	defaultProps: {
+		direction: 'down',
+		open: false
+	},
+
+	styles: {
+		css: componentCss,
+		className: 'Dropdown',
```
Comment (teckliew) [inline Dropdown/Dropdown.js]: Lowercase for css class name: `dropdown`

File: Dropdown/Dropdown.js [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+		 *
+		 * @type {Function}
+		 */
+		onSelect: PropTypes.func,
+		/*
+		 * Called when clicked on the dropdown to open.
+		 *
+		 * @type {Number}
+		 */
+
```
Comment (teckliew) [inline Dropdown/Dropdown.js]: Spacing. Also there should be a line between each prop type block.

File: Dropdown/Dropdown.js [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+		/**
+		 * Customizes the component by mapping the supplied collection of CSS class names to the
+		 * corresponding internal elements and states of this component.
+		 *
+		 * @type {Object}
+		 * @public
+		 */
+		css: PropTypes.object,
+		/**
+		 * The animation direction of the `palette`.
```
Comment (teckliew) [inline Dropdown/Dropdown.js]: Maybe update the doc to "the direction where the dropdown list appears"?

File: Dropdown/Dropdown.js [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+
+	styles: {
+		css: componentCss,
+		className: 'Dropdown',
+		publicClassNames: true
+	},
+
+	computed: {
+		className: ({css, selected, styler}) => styler.append(selected && css.selected),
+		transitionContainerClassname: ({css, open, styler}) => styler.join(css.transitionContainer, (open ? css.openTransitionContainer : null)),
```
Comment (teckliew) [inline Dropdown/Dropdown.js]: We could clean this up to just be `styler.join(css.transitionContainer, {css.openTransitionContainer : open})`

File: Dropdown/Dropdown.js [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+				</Transition>
+			</div>
+		);
+	}
+});
+
+
+
+/**
+ * Applies Agate specific behaviors to [ColorPickerBase]{@link agate/Dropdown.DropdownBase}.
```
Comment (teckliew) [inline Dropdown/Dropdown.js]: `ColorPickerBase` should be updated to `DropdownBase`

File: Dropdown/Dropdown.module.less [hunk 7]
```diff
@@ -0,0 +1,39 @@
+// Dropdown.module.less
+//
+@import "../styles/mixins.less";
+@import "../styles/skin.less";
+
+.Dropdown {
+	.applySkins({
+		.item {
```
Comment (teckliew) [inline Dropdown/Dropdown.module.less]: Do we need to override the item styles or should we just keep the default item styles?

File: Dropdown/Dropdown.module.less [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+			border-color: gray;
+			border-style: solid;
+			border-width: 1px;
+		}
+		.transitionContainer {
+			position: absolute;
+			display: inline-block;
+			z-index: 1;
+			pointer-events: none;
+			min-width: 312px;
```
Comment (teckliew) [inline Dropdown/Dropdown.module.less]: We should probably have this as a less variable, along with all the other properties like the background color and border styles in `.dropdownList`, and the height in `.dropdownListWithScroller`

File: Dropdown/Dropdown.module.less [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+			width: 100%;
```
Comment (teckliew) [inline Dropdown/Dropdown.module.less]: Should we keep this to 100%? Or how should we support different dropdown widths?

Comment (RenukaAtale) [reply Dropdown/Dropdown.module.less]: At this point we need keep the width,  don't have enough information for different sizes for dropdown.

File: Dropdown/Dropdown.module.less (hunk 7 above)
Comment (RenukaAtale) [reply Dropdown/Dropdown.module.less]: I have added the style on the first item for border. we won't be able to use the default item style as don't have default border on the item.

File: Dropdown/Dropdown.js [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+	computed: {
+		transitionContainerClassname: ({css, open, direction, styler}) => styler.join(css.transitionContainer, {openTransitionContainer: open, upTransitionContainer: direction === 'up'} ),
+		dropdownListClassname: ({children, css, styler}) => styler.join(css.dropdownList, {dropdownListWithScroller: children.length > 4}),
+		title: ({children, selected}) => {
+			if (isSelectedValid({children, selected})) {
+				const child = children[selected];
+				return typeof child === 'object' ? child.children : child;
+			}
+
+			return children[0];
```
Comment (webOS101) [inline Dropdown/Dropdown.js]: `children` is not required.  Will this throw if `children` is `null`?

File: Dropdown/Dropdown.js [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+		 */
+		css: PropTypes.object,
+
+		/**
+		 * The direction where the dropdown list appears.
+		 *
+		 * @type {String}
+		 * @public
+		 */
+		direction: PropTypes.string,
```
Comment (webOS101) [inline Dropdown/Dropdown.js]: Can you enumerate the valid values here and in the jsdoc?

File: Dropdown/Dropdown.js [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+ * @memberof agate/Dropdown
+ * @ui
+ * @public
+ */
+const DropdownBase = kind({
+	name: 'Dropdown',
+
+	propTypes: /** @lends agate/Dropdown.Dropdown.prototype */ {
+		/*
+		 * The selections for Dropdown
```
Comment (webOS101) [inline Dropdown/Dropdown.js]: Should we mention that the title will be the first item if no items are selected (which seems odd to me, is that per the design)?

File: Dropdown/Dropdown.js [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+
+		/*
+		 * Called when an item is selected.
+		 *
+		 * @type {Function}
+		 */
+		onSelect: PropTypes.func,
+
+		/*
+		 * Called when clicked on the dropdown to open.
```
Comment (webOS101) [inline Dropdown/Dropdown.js]: This is not a callback, but the open state.  Add the `@default` value here?

File: Dropdown/Dropdown.js [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+		transitionDirection: ({direction}) => {
+			switch (direction) {
+				case 'left':
+					return 'right';
+				case 'right':
+					return 'left';
+				case 'up':
+					return 'down';
+				case 'down':
+					return 'up';
```
Comment (webOS101) [inline Dropdown/Dropdown.js]: Should we add `default:` here?

File: Dropdown/Dropdown.js [hunk 15]
```diff
@@ -0,0 +1,200 @@
...
+							>{children || []}</Group>
+						</Scroller>
+					</ContainerDiv>
+				</Transition>
+			</div>
+		);
+	}
+});
+
+
```
Comment (webOS101) [inline Dropdown/Dropdown.js]: Can delete these extra empty line here and below.

File: Dropdown/Dropdown.js [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+ *
+ * By default, `Dropdown` maintains the state of its `selected` property. Supply the
+ * `defaultSelected` property to control its initial value. If you wish to directly control updates
+ * to the component, supply a value to `selected` at creation time and update it in response to
+ * `onSelected` events.
+ *
+ * @class Dropdown
+ * @memberof agate/Dropdown
+ * @extends agate/Dropdown.DropdownBase
+ * @ui
```
Comment (webOS101) [inline Dropdown/Dropdown.js]: `@mixes agate/Dropdown.DropdownDecorator`

File: Dropdown/Dropdown.js [hunk 17]
```diff
@@ -0,0 +1,200 @@
...
+		 *
+		 * @type {String[]|Array.<{key: (Number|String), children: (String|Component)}>}
+		 */
+		children: PropTypes.array,
+
+		/**
+		 * Customizes the component by mapping the supplied collection of CSS class names to the
+		 * corresponding internal elements and states of this component.
+		 *
+		 * @type {Object}
```
Comment (webOS101) [inline Dropdown/Dropdown.js]: Are there any customizable classes?  If not, maybe we should just mark this private?  If there are, can you document them here?

File: Dropdown/Dropdown.js [hunk 18]
```diff
@@ -0,0 +1,200 @@
...
+ *
+ * @class DropdownBase
+ * @memberof agate/Dropdown
+ * @ui
+ * @public
+ */
+const DropdownBase = kind({
+	name: 'Dropdown',
+
+	propTypes: /** @lends agate/Dropdown.Dropdown.prototype */ {
```
Comment (webOS101) [inline Dropdown/Dropdown.js]: `DropdownBase.prototype`

File: Dropdown/Dropdown.js (hunk 14 above)
Comment (webOS101) [reply Dropdown/Dropdown.js]: I think this is the opposite of what we would want, isn't it?  The transition is opposite from the opening so it should be the same as the default.

File: Dropdown/Dropdown.js [hunk 19]
```diff
@@ -0,0 +1,224 @@
...
+ * @class DropdownBase
+ * @memberof agate/Dropdown
+ * @ui
+ * @public
+ */
+const DropdownBase = kind({
+	name: 'Dropdown',
+
+	propTypes: /** @lends agate/Dropdown.DropdownBase.prototype */ {
+		/*
```
Comment (webOS101) [inline Dropdown/Dropdown.js]: jsdoc comments need to start with `/**` or else they won't be parsed.  I have fixed these.

File: Dropdown/Dropdown.js [hunk 20]
```diff
@@ -0,0 +1,222 @@
+/**
+ * Agate styled Header.
```
Comment (webOS101) [inline Dropdown/Dropdown.js]: oops.

**Pull request merged**